### PR TITLE
perf(gemm): split-K, occupancy, and m=1 stream-GEMV decode tactics for mm_bf16_fp4

### PR DESCRIPTION
**Pull request by yichengj0**
## 📌 Description

#3597 added `mm_bf16_fp4` (bf16 activations x nvfp4 weights) so vLLM can replace Marlin, its default W4A16 backend, with a FlashInfer kernel; DGX Spark is the lead target. On decode shapes the kernel trailed Marlin for two reasons: small-n grids underfill the GPU, and at single-token batches (m=1) too few resident warps per SM hide DRAM latency. This PR addresses both: m=1 decode beats or matches Marlin on every part we measured, and end-to-end serving of Qwen3.6-27B-NVFP4 flips from behind to ahead at batch 1 with no regression elsewhere.

**What it does**

- Adds split-K tactics (2/4/8 splits) to the autotuner space, offered only when splitting shortens the grid's last wave by at least 25%. Splits write fp32 partials, and a PDL-chained reduce kernel sums them in fixed order, so results are deterministic run to run.
- Adds occupancy tactics: 2 or 3 co-resident CTAs per SM trade pipeline depth for latency hiding on weight-bound grids, plus occupancy 2 combined with split-K for narrow-n shapes.
- Adds a streaming GEMV kernel (`gemv_bf16_fp4_sm12x.py`) for the bandwidth-bound m=1 case: no shared memory, no tensor cores, weights stream from global memory to registers with latency hidden by warp count. It reads the same packed operands as the MMA kernel, so the autotuner picks between the two per shape.
- Sizes GEMV split-K from the device: alongside the power-of-2 splits, the menu carries a split targeting ~20 warps/SM (the measured saturation point). Tactic indices become device-scoped, so the autotuner cache key now carries the SM count.
- Routes the no-autotune m=1 fallback onto the GEMV. Serving stacks do not tune every shape (vLLM's warmup never captures the logits GEMM), so the lm_head always takes this path.
- Fixes kernel launch to pass no cluster dimensions: the boilerplate `cluster=[1,1,1]` routed launches through the cluster work distributor, whose co-residency cap silently defeated the occupancy tactics on SM12x.

No public API changes. The one observable behavior change: untuned m=1 calls on SM12x now run the GEMV, whose output is bitwise different from the MMA heuristic's but equally accurate and still deterministic.

#### Split-K on the #3597 decode shapes (RTX PRO 6000 / DGX Spark / RTX 5080)

Single-token decode GEMMs (m=1). The first table covers serving-class shapes (Llama-8B projection layers plus the #3597 example shape); the second covers #3597's own benchmark grid. Median GPU time over CUDA-graph replays with a cold L2 cache, as in serving. Baseline is vLLM's Marlin on the same GPU; FlashInfer runs with autotuning. Speedup = Marlin time / FlashInfer time. Each cell is a RTX PRO 6000 / DGX Spark / RTX 5080 triple.

| n x k | vs Marlin, before this PR | vs Marlin, with this PR |
|--:|:--:|:--:|
| 2048x7168 | 0.57 / 1.00 / 0.47 | **1.36** / **1.02** / **0.78** |
| 4096x4096 | 0.75 / 0.91 / 0.90 | **1.06** / **1.00** / 0.90 |
| 4096x14336 | 0.60 / 0.91 / 0.78 | **0.98** / **1.00** / 0.78 |
| 14336x4096 | 0.97 / 1.00 / 0.86 | 0.97 / 1.00 / 0.86 |
| 10304x2688 | 0.78 / 0.98 / 0.96 | 0.78 / 0.98 / 0.96 |

The same comparison over #3597's benchmark grid (4096x4096 appears in the table above):

| n x k | vs Marlin, before this PR | vs Marlin, with this PR |
|--:|:--:|:--:|
| 512x2048 | 2.09 / 0.95 / 0.95 | **4.48** / **1.43** / **2.30** |
| 512x4096 | 1.19 / 0.70 / 0.56 | **3.61** / **1.26** / **1.88** |
| 1024x2048 | 1.26 / 0.98 / 0.70 | **2.39** / **1.03** / **1.32** |
| 1024x4096 | 0.99 / 0.86 / 0.49 | **2.72** / **1.04** / **1.21** |
| 2048x512 | 1.86 / 1.25 / 1.19 | 1.86 / 1.25 / 1.19 |
| 2048x1024 | 1.36 / 1.16 / 0.90 | **1.40** / 1.16 / **1.02** |
| 2048x2048 | 1.17 / 1.11 / 0.70 | **1.72** / **1.02**\* (1.11) / **0.97** |
| 2048x4096 | 0.73 / 1.08 / 0.54 | **1.48** / **1.04**\* (1.08) / **0.84** |
| 4096x512 | 1.40 / 1.01 / 1.38 | 1.40 / **0.95**\* (1.01) / 1.38 |
| 4096x1024 | 1.41 / 0.93 / 1.18 | 1.41 / **0.95** / 1.18 |
| 4096x2048 | 0.95 / 0.95 / 1.06 | 0.95 / **1.00** / 1.06 |
| 131072x2048 | 0.98 / 0.97 / 0.90 | 0.98 / 0.97 / 0.90 |
| 248320x2048 | 0.98 / 1.00 / 0.93 | 0.98 / 1.00 / 0.93 |

- Bold marks the cells this PR changes (the tuner picks a new split-K tactic); unbolded picks perform as before.
- \* These three cells are tuner mis-picks, not kernel regressions: an accurate pick would keep #3597's pre-existing non-split config, and the value in parentheses is what that config achieves. The Reviewer Notes explain the cause.
- On the serving shapes, the RTX 5080 column stays below 1.0 even where this PR helps. Profiling of the larger losses points to activation re-reads through L2, a separate problem from grid fill and out of scope here.

#### Qwen3.6-27B-NVFP4 decode GEMMs at m=1 (RTX 5080, DGX Spark)

Same methodology as above; speedup = Marlin time / FlashInfer time.

| GEMM (n x k) | RTX 5080 | DGX Spark |
|--|:--:|:--:|
| gate_up 34816x5120 | **1.03** | 1.00 |
| down 5120x17408 | **1.03** | 1.00 |
| lm_head 248320x5120 | **1.04**\* | **1.04** |

\* Marlin's lm_head repack does not fit on the 16 GB RTX 5080, so this cell compares against the best in-tree MMA tactic. Spark ties at its bandwidth floor on the first two shapes.

#### End-to-end vLLM serving of Qwen3.6-27B-NVFP4 (RTX PRO 6000)

Full serving A/B, FlashInfer leg vs Marlin leg under identical settings, aiperf output-token throughput, ratio = FlashInfer / Marlin: batch-1 decode 1.011x, low-concurrency speculative decode 1.08 to 1.11x ahead, and every other cell at parity within run-to-run noise (0.978 to 1.014x) with no regression beyond it.

## 🔍 Related Issues

Follow-up to #3597.

## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [x] I have installed the hooks with `pre-commit install`.
- [x] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

## 🧪 Tests

- [x] Tests have been added or updated as needed.
- [x] All tests are passing (`unittest`, etc.).

New tests:

- Every enumerated tactic (MMA and GEMV) is checked against a reference and for bit-exact run-to-run determinism.
- Unit tests pin the fallback selectors' picks; one test drives tactic=-1 through the GEMV fallback end to end.
- Full test file passes on RTX 5080, RTX PRO 6000, and GB10.

## Reviewer Notes

- Most gains require autotuning, which serving frameworks run at startup. The no-autotune fallback picks match the tuner's choices on every part we measured.
- The autotuner times candidates with a warm L2 while decode serving runs cold, so it can over-rank split tactics; the 25% last-wave guard compensates but does not fully close it (the three Spark cells in the grid table). This measurement gap is general and deserves its own issue.
- The fallback picks add JIT-compiled kernel variants per decode shape class, cached in-process only; that cost amortizes to once per machine when this module adopts the #3874 CuTe-DSL disk cache, as #4029 did for the sibling `mm_fp4` path. The GEMV's device-derived splits widen this surface, so the follow-up is worth prioritizing.
- Other FlashInfer cute-dsl kernels also pass `cluster=[1,1,1]` at launch and inherit the same co-residency cap; they are worth a separate audit.

## Summary by CodeRabbit

- **New Features**
  - Added split-K support for bf16 × fp4 matrix multiplication to improve performance across varying workloads.
  - Added a dedicated SM12x GEMV path for efficient single-row operations.
  - Added automatic tuning for split counts, occupancy, and device-specific execution strategies.
  - Added support for FP16 GEMV outputs and deterministic partial-result reduction.

- **Bug Fixes**
  - Improved handling of GEMV and split-K fallback selection across supported shapes and GPU configurations.

- **Tests**
  - Added coverage for accuracy, determinism, GEMV correctness, and split-K selection.

### CODE REVIEW

File: tests/gemm/test_mm_bf16_fp4.py [hunk 1]
```diff
@@ -270,6 +271,183 @@ def test_backend_out_dtype_override(backend):
...
+    ref = (a.float() @ weight_fp32.T).to(torch.bfloat16)
+
+    runner = _cute_dsl_bf16_fp4_runner(enable_pdl=True)
+    sf_u8 = sf_p.view(torch.uint8).contiguous()
+    alpha_l = _prepare_bf16_fp4_alpha(alpha_p, device)
+    for tactic, cfg in enumerate(
+        _bf16_fp4_cute_dsl_tactic_configs(n, k, get_device_sm_count(device))
+    ):
+        if cfg[7] == "gemv" and m != 1:
+            continue
```
[the remark this answers is not in the document]
Comment (yichengj0) [reply tests/gemm/test_mm_bf16_fp4.py:305]: Fixed in c806b5e6: the every-tactic loop now skips gemv configs off SM12x, and the fp16-out test skips up front. The tactic=-1 fallback test was already covered by its selector guard.

**Comment by bkryu** [comment]
@flashinfer-bot run

**Review by bkryu — APPROVED** [review]
